File: examples/producer/rust/src/producer/responses/message_request.rs
use super::{identification, producer::Control, protocol, scope::Scope, Context};
use clibri::server;
use uuid::Uuid;

type BroadcastEventsMessage = (Vec<Uuid>, protocol::Events::Message);

pub enum Response {
    Accept((protocol::Message::Accepted, BroadcastEventsMessage)),
    Deny(protocol::Message::Denied),
}

#[allow(unused_variables)]
pub async fn response<E: server::Error, C: server::Control<E>>(
    request: &protocol::Message::Request,
    scope: &mut Scope<'_, E, C>,
) -> Result<Response, protocol::Message::Err> {
    let uuid = scope.identification.uuid();
    if !scope.context.is_user_exist(&request.user).await {
        Ok(Response::Deny(protocol::Message::Denied {
            reason: String::from("User doesn't exist"),
        }))
    } else {
        match scope
            .context
            .add_message(&request.user, request.message.clone())
        {
            Ok(msg) => Ok(Response::Accept((
                protocol::Message::Accepted {
                    uuid: msg.uuid.clone(),
                },
                (scope.filter.except(&uuid), msg),
            ))),
            Err(err) => Err(protocol::Message::Err { error: err }),
        }
    }
}

File: examples/producer/typescript/src/producer/responses/message.request.ts
import { Response } from "../implementation/responses/message.request";
import {
	Context,
	Producer,
	Identification,
	Filter,
	Protocol,
} from "../implementation/responses";
import { Scope } from "../implementation/scope";

export function response(
	request: Protocol.Message.Request,
	scope: Scope
): Promise<Response> {
	const msg = scope.context.addMessage(
		new Protocol.Messages.Message({
			user: request.user,
			uuid: scope.consumer.uuid(),
			message: request.message,
			timestamp: BigInt(Date.now()),
		})
	);
	return Promise.resolve(
		new Response(
			new Protocol.Message.Accepted({
				uuid: scope.consumer.uuid(),
			})
		)
			.broadcast(scope.filter.except(scope.consumer.uuid()))
			.EventsMessage(
				new Protocol.Events.Message({
					uuid: msg.uuid,
					user: msg.user,
					message: msg.message,
					timestamp: msg.timestamp,
				})
			)
	);
}
